src/adapter/src/coord.rs

```diff
@@ -555,35 +555,162 @@ impl PendingReadTxn {
+/// through various functions as part of statement execution.
+/// Currently, it is only used to finalize execution, by calling one
+/// of the methods `retire` or `retire_aysnc`. Finalizing execution
+/// involves sending the session back to the pgwire layer so that it
+/// may be used to process further commands. In the future, it will
+/// also involve performing some work on the main coordinator thread
+/// (e.g., recording the time at which the statement finished
+/// executing) the state necessary to perform this work is bundled in
+/// the `ExecuteContextExtra` object (today, it is simply empty).
+#[derive(Debug)]
+pub struct ExecuteContext {
+    tx: ClientTransmitter<ExecuteResponse>,
+    internal_cmd_tx: mpsc::UnboundedSender<Message>,
+    session: Session,
+    extra: ExecuteContextExtra,
+}
+
+impl ExecuteContext {
+    pub fn session(&self) -> &Session {
+        &self.session
+    }
+
+    pub fn session_mut(&mut self) -> &mut Session {
+        &mut self.session
+    }
+
+    pub fn tx(&self) -> &ClientTransmitter<ExecuteResponse> {
+        &self.tx
+    }
+
+    pub fn tx_mut(&mut self) -> &mut ClientTransmitter<ExecuteResponse> {
+        &mut self.tx
+    }
+
+    pub fn from_parts(
+        tx: ClientTransmitter<ExecuteResponse>,
+        internal_cmd_tx: mpsc::UnboundedSender<Message>,
+        session: Session,
+        extra: ExecuteContextExtra,
+    ) -> Self {
+        Self {
+            tx,
+            session,
+            extra,
+            internal_cmd_tx,
+        }
+    }
+
+    /// By calling this function, the caller takes responsibility for
+    /// dealing with the instance of `ExecuteContextExtra`. This is
+    /// intended to support protocols (like `COPY FROM`) that involve
+    /// multiple passes of sending the session back and forth between
+    /// the coordinator and the pgwire layer. As part of any such
+    /// protocol, we must ensure that the `ExecuteContextExtra`
+    /// (possibly wrapped in a new `ExecuteContext`) is passed back to the coordinator for
+    /// eventual retirement.
+    pub fn into_parts(
+        self,
+    ) -> (
+        ClientTransmitter<ExecuteResponse>,
+        mpsc::UnboundedSender<Message>,
+        Session,
+        ExecuteContextExtra,
+    ) {
+        let Self {
+            tx,
+            internal_cmd_tx,
+            session,
+            extra,
+        } = self;
+        (tx, internal_cmd_tx, session, extra)
+    }
+
+    /// Retire the execution, with access to the coordinator.
+    ///
+    /// Since the caller provides us with exclusive access to the coordinator, we are
+    /// able to perform whatever action necessary synchronously on the main coordinator task.
+    ///
+    /// If you find yourself in a context where you don't have exclusive access to the coordinator,
+    /// use `retire_aysnc` instead.
+    pub fn retire(self, result: Result<ExecuteResponse, AdapterError>, coord: &mut Coordinator) {
```

Comment:
I'm not convinced of the need to have both this and retire_async. It's requiring the caller to know if it wants its side effects processed now or enqueued. I don't think the caller should care about that. If we do need to know this, it should perhaps be a method of ContextExtra, then we can only have one retire function.

Reply:
We could have `retire_async` be the only option (and probably just rename it to `retire`). It just seemed slightly wasteful to enqueue a whole new message and wait for the controller to get to it, rather than to process it immediately, in the cases where we _do_ have `&mut` access to the controller at the end of the statement lifecycle (which is the vast majority of cases).

My intention here was that call sites that are running on the main coordinator task (and thus have `&mut self`) will always call `retire`, and other call sites will always call `retire_async`. I can try to document that idea better, or I can just get rid of the two different functions and always enqueue a message. I'll leave it up to you and @jkosh44 since your team owns this part of the code.

Reply:
I think there is an argument to be made that statement logging is high priority, because in some pathological case where the coordinator is being hammered with commands, you might be particularly interested in investigating what statements are being executed, for debugging purposes.

However, I admit that's a rather niche case, and probably adequately served by just having a simple QPS prometheus metric. So I'm leaning towards your suggestion of a low-pri channel.

Reply:
How you've written it currently, to use `internal_cmd_tx`, might be a nice middle ground. It would yield to allow any other messages on the same channel to be executed, if any exist, but would never get starved out.

I don't have a strong opinion between the proposed options, though I'm leaning slightly to one of the options where we always send a message because it's a bit simpler on the caller. Also it seems to be an existing pattern in the Coordinator to prefer sending messages when it's ambiguous whether or not we're on the main Coord thread. For example,  https://github.com/MaterializeInc/materialize/blob/0bce7396904d49d4d8eff62d98371b908ef7c9d2/src/adapter/src/util.rs#L78-L92